core/src/main/scala/org/apache/spark/Accumulable.scala

```diff
@@ -35,40 +35,67 @@ import org.apache.spark.util.Utils
  * [[org.apache.spark.Accumulator]]. They won't always be the same, though -- e.g., imagine you are
  * accumulating a set. You will add items to the set, and you will union two sets together.
  *
+ * All accumulators created on the driver to be used on the executors must be registered with
+ * [[Accumulators]]. This is already done automatically for accumulators created by the user.
+ * Internal accumulators must be explicitly registered by the caller.
+ *
+ * Operations are not thread-safe.
+ *
+ * @param id ID of this accumulator; for internal use only.
  * @param initialValue initial value of accumulator
  * @param param helper object defining how to add elements of type `R` and `T`
  * @param name human-readable name for use in Spark's web UI
  * @param internal if this [[Accumulable]] is internal. Internal [[Accumulable]]s will be reported
  *                 to the driver via heartbeats. For internal [[Accumulable]]s, `R` must be
  *                 thread safe so that they can be reported correctly.
+ * @param countFailedValues whether to accumulate values from failed tasks. This is set to true
```

Comment:
This reminds me: I believe that there's a paragraph in the Spark Programming Guide which describes the current accumulator semantics (sorta); it would be good to add new documentation for theses metric-style accumulators, since I imagine that a number of users would be interested in taking advantage of this new opt-in semantic in their own code.

We shouldn't block merging this patch on those doc updates (since I'm worried about merge-conflict potential and want to get this in sooner than later), but let's file a followup ticket so we don't forget.